okteto/okteto
# Always pull & export default cache when building an image

# Proposed changes

Fixes #3426

## How to test (Okteto Cloud)

1. `git clone git@github.com:okteto/movies-frontend.git`
2. Edit the `okteto.yml` to be like:
```YAML
build:
  frontend:
    context: .
    dockerfile: Dockerfile
    image: okteto.dev/frontend:1.0.0
    export_cache: okteto.dev/frontend:cache

deploy:
  - name: Deploy Frontend
    command: helm upgrade --install movies-frontend chart --set image=${OKTETO_BUILD_FRONTEND_IMAGE}

dev:
  frontend:
    image: ${OKTETO_BUILD_FRONTEND_DEV_IMAGE}
    command: yarn start
    sync:
      - .:/src
```
3. Run `okd build -l=debug`
4. Notice how you get both errors on pulling the cache images (like:  => `ERROR importing cache manifest from registry.cloud.okteto.net/okteto/frontend:cache` for both `/okteto` and `/{your-username`)
5. When the build suceeds it will export the cache to `okteto.dev`, now run again the build command 
6. Notice how this time the cache from `/{your-username}` is pulled successfully with a message like ` => importing cache manifest from registry.cloud.okteto.net/{your-username}/frontend:cache`

## How to test (Dev Cluster)

1. Switch to your dev cluster context
2. Follow the same steps for testing Okteto Cloud with the difference that here you should be able to push and pull the cache from the Global Registry

## How to test (Vanilla Cluster)

1. Using Docker Desktop, turn on the Kubernetes cluster
4. Select that okteto context (ie. `okteto context use docker-desktop`)
5. Run `okd build -l=debug`
6. Notice how you don't see any errors of cache fail being pulled either being imported
